app/assets/javascripts/shared/datatable/main.js

```diff
@@ -0,0 +1,199 @@
+            name: 'mergeBtn',
+            className: 'd-none',
+            action: this.mergeSelected.bind(this)
+          },
+          {
+            autoClose: true,
+            available: function(){
+              return that.$table.data('tags') !== undefined;
+            },
+            className: 'd-none',
+            extend: 'collection',
+            name: 'tagBtn',
+            text: '<i class="fa fa-tags fa-fw"></i>Tag<i class="fa fa-caret-down fa-fw"></i>',
+            buttons: this.setupTagButtons()
+          },
+          {
+            extend: 'colvis',
+            text: '<i class="fa fa-columns mr-1"></i><i class="fa fa-caret-down"></i>',
+            titleAttr: 'Choose columns to show',
+            className: 'btn',
+            columns: columnVisibleIndexes
+          }
+        ]
+      },
+      columnDefs: [
+        {
+          targets: hiddenColumnIndexes,
+          visible: false
+        }
+      ],
+      dom: "<'row'<'col-lg-6'B><'col-lg-6'f>>" +
+        "<'row'<'col-lg-12'tr>>" +
+        "<'dataTables_footer_content'ip>",
+      initComplete: function (settings) {
+        settings.oInstance.wrap("<div class='table-wrapper'></div>");
+      },
+      lengthChange: false,
+      pageLength: 25,
+      stateSave: true,
+      stateSaveCallback: function(settings, data) {
+        localStorage.setItem(that.localStorageKey, JSON.stringify(data));
+      },
+      stateLoadCallback: function(settings) {
+        return JSON.parse(localStorage.getItem(that.localStorageKey));
+      },
+      select: {
+        selector: 'td.select-checkbox',
+        style: 'multi'
+      }
+    });
+
+    this.validateRecords();
+
+    this.behaviors();
+  }
+
+  behaviors() {
+    this.setupCheckboxListeners();
+    this.setupMergeButtonToggle();
+    this.setupTagButtonToggle();
+    this.setupBulkDeleteButtonToggle();
+
+    this.unbindDataTable();
+  }
+
+  toggleLoadingState(rows, isLoading, buttonName) {
+    var button = this.dataTable.buttons(buttonName + ':name');
+
+    $(button[0].node).toggleClass('disabled', isLoading);
+
+    rows.nodes().toArray().forEach(function(tr) {
+      if (isLoading) {
+        $(tr).find('[data-behavior~=error-loading]').remove();
+        $(tr).find('[data-behavior~=select-checkbox]').append('<div class="spinner-border spinner-border-sm text-primary" data-behavior="spinner"><span class="sr-only">Loading</div>');
+      } else {
+        $(tr).find('[data-behavior~=spinner]').remove();
+      }
+    })
+  }
+
+  rowIds(rows) {
+    var ids = rows.ids().toArray().map(function(id) {
+      // The dom id for <tr> is in the following format: <tr id="item_name-id"></tr>,
+      // so we split it by the delimiter to get the id number.
+      return id.split('-')[1];
+    });
+    return ids;
+  }
+
+  unbindDataTable() {
+    var that = this;
+
+    document.addEventListener('turbolinks:before-cache', function() {
+      that.dataTable.destroy();
+    });
+  }
+
+  validateRecords() {
```

Comment:
Can we move this into a file as well? It'll shorten the file by 30 lines :smile: